# **04: MongoDB PyMongo and Query Optimisation**

**Module:** Databases and Analytics (CP6UA56O)  
**Case study:** NorthStar Urban Mobility and Logistics

This notebook combines the MongoDB Atlas implementation and MongoDB query optimisation sections. Cleaned NorthStar data is remodelled into MongoDB collections using PyMongo, then important business queries are optimised using indexes and explain plans.

# **Section 1 - MongoDB Design & Implementation (Data Insertion)**

This section connects to MongoDB Atlas, creates the NorthStar database structure, inserts cleaned data into collections, and validates the inserted document counts. The main operational collection is `service_cases`, which is designed around the service case view of NorthStar's orders, deliveries, complaints, incidents, app events, and operational analytics.

This section also implements a NoSQL schema suited to NorthStar's operational reporting needs. Master collections store entities such as customers, drivers, vehicles, hubs, and app events, while `service_cases` embeds related order, delivery, complaint, incident, analytics, and status flag information for efficient case-level querying.

---
## **Install and Import Required Libraries**

`pymongo` is used to connect Python with MongoDB Atlas. `pandas` and `numpy` are used to prepare cleaned CSV data before converting it into MongoDB documents.

In [ ]:
!pip -q install pymongo dnspython

In [ ]:
import pandas as pd
import numpy as np
from pymongo import MongoClient, ASCENDING, DESCENDING
from pymongo.errors import ServerSelectionTimeoutError, OperationFailure, BulkWriteError
from pprint import pprint
from pathlib import Path
import json
import re
import shutil
import copy
import certifi
from bson import json_util

pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', 120)

print('Libraries loaded successfully.')

Libraries loaded successfully.


---
## **Load Cleaned Data from GitHub**

The cleaned files were produced in Notebook 1 and uploaded to the GitHub repository under `Data/Cleaned`. This notebook imports those files directly from GitHub into Colab. This keeps the workflow aligned with the coursework requirement that the dataset is stored in GitHub and imported into Google Colab.

In [ ]:
CLEANED_BASE_URL = 'https://raw.githubusercontent.com/SamsonSiby5827/Northstar_Databases_Analytics_Coursework/main/Data/Cleaned'

def github_csv(filename):
    url = f'{CLEANED_BASE_URL}/{filename}'
    df = pd.read_csv(url)
    print(f'{filename:<35} {df.shape[0]} rows, {df.shape[1]} columns')
    return df

customers   = github_csv('customers_clean.csv')
orders      = github_csv('orders_clean.csv')
deliveries  = github_csv('deliveries_clean.csv')
drivers     = github_csv('drivers_clean.csv')
vehicles    = github_csv('vehicles_clean.csv')
hubs        = github_csv('hubs_clean.csv')
complaints  = github_csv('complaints_clean.csv')
incidents   = github_csv('incidents_clean.csv')
app_events  = github_csv('app_events_clean.csv')
ops_df      = github_csv('ops_merged.csv')
complaint_df = github_csv('complaints_merged.csv')
incident_df  = github_csv('incidents_merged.csv')

print('\nAll cleaned datasets were loaded from GitHub.')

customers_clean.csv                 650 rows, 10 columns
orders_clean.csv                    1250 rows, 16 columns
deliveries_clean.csv                950 rows, 21 columns
drivers_clean.csv                   170 rows, 9 columns
vehicles_clean.csv                  120 rows, 11 columns
hubs_clean.csv                      8 rows, 6 columns
complaints_clean.csv                320 rows, 11 columns
incidents_clean.csv                 280 rows, 7 columns
app_events_clean.csv                640 rows, 12 columns
ops_merged.csv                      1250 rows, 59 columns
complaints_merged.csv               320 rows, 22 columns
incidents_merged.csv                280 rows, 20 columns

All cleaned datasets were loaded from GitHub.


---
## **Prepare Date and Numeric Columns**

Before creating MongoDB documents, important date and numeric fields are converted into suitable formats. This helps MongoDB store values more consistently and avoids passing unclear Pandas types into BSON documents.

In [ ]:
# Convert common date columns where they exist.
date_map = [
    (orders, ['order_created_at']),
    (deliveries, ['dispatch_time', 'delivery_completed_at']),
    (complaints, ['created_at', 'order_created_at']),
    (incidents, ['reported_at']),
    (app_events, ['event_timestamp']),
    (ops_df, ['order_created_at', 'dispatch_time', 'delivery_completed_at']),
    (complaint_df, ['created_at', 'order_created_at']),
    (incident_df, ['reported_at'])
]

for df, cols in date_map:
    for col in cols:
        if col in df.columns:
            df[col] = pd.to_datetime(df[col], errors='coerce')

# Convert important numeric columns where they exist.
numeric_cols = [
    'order_value', 'promised_window_hours', 'route_distance_km',
    'manual_route_override_count', 'customer_rating_post_delivery',
    'fuel_or_charge_cost', 'completion_hours', 'completion_hours_clean',
    'cost_per_km', 'negative_completion_time_flag', 'missed_promised_window',
    'has_complaint', 'has_incident', 'is_bad_outcome',
    'resolution_days', 'compensation_amount',
    'api_latency_ms', 'success_flag', 'loyalty_score', 'years_experience',
    'training_score', 'driver_rating', 'battery_health_pct',
    'low_battery_flag', 'capacity_score'
]

all_dataframes = [customers, orders, deliveries, drivers, vehicles, hubs,
                  complaints, incidents, app_events, ops_df, complaint_df, incident_df]

for df in all_dataframes:
    for col in numeric_cols:
        if col in df.columns:
            df[col] = pd.to_numeric(df[col], errors='coerce')

print('Date and numeric preparation complete.')

load_summary = pd.DataFrame({
    'dataset': ['customers', 'orders', 'deliveries', 'drivers', 'vehicles', 'hubs',
                'complaints', 'incidents', 'app_events', 'ops_df', 'complaint_df', 'incident_df'],
    'rows': [len(customers), len(orders), len(deliveries), len(drivers), len(vehicles), len(hubs),
             len(complaints), len(incidents), len(app_events), len(ops_df), len(complaint_df), len(incident_df)],
    'columns': [customers.shape[1], orders.shape[1], deliveries.shape[1], drivers.shape[1], vehicles.shape[1],
                hubs.shape[1], complaints.shape[1], incidents.shape[1], app_events.shape[1],
                ops_df.shape[1], complaint_df.shape[1], incident_df.shape[1]]
})

load_summary

Date and numeric preparation complete.


,dataset,rows,columns
0,customers,650,10
1,orders,1250,16
2,deliveries,950,21
3,drivers,170,9
4,vehicles,120,11
5,hubs,8,6
6,complaints,320,11
7,incidents,280,7
8,app_events,640,12
9,ops_df,1250,59


---
## **MongoDB Schema Design**

The MongoDB design uses a combination of master collections and one main nested collection.

### Master collections

These collections store reusable records:

- `customers`
- `drivers`
- `vehicles`
- `hubs`
- `app_events`

### Main analytical collection

The main collection is `service_cases`. Each document is based on one order and embeds related operational records such as delivery, complaints, incidents, and order-linked app events.

### Design choice

Embedding is used for complaints, incidents, and app events because these records describe the history of a specific service case. Master records are also stored separately because customers, drivers, vehicles, and hubs can be referenced across many cases. This design supports business queries while still preserving a clear NoSQL document structure.

---
## **Helper Functions for MongoDB Documents**

MongoDB stores documents in BSON format. Pandas and NumPy values such as `NaN`, `Timestamp`, and NumPy numbers need to be converted into normal Python values before insertion.

In [ ]:
def clean_value(value):
    # Convert Pandas and NumPy scalar values into MongoDB friendly values.
    if pd.isna(value):
        return None
    if isinstance(value, pd.Timestamp):
        return value.to_pydatetime()
    if isinstance(value, np.datetime64):
        return pd.to_datetime(value).to_pydatetime()
    if isinstance(value, np.integer):
        return int(value)
    if isinstance(value, np.floating):
        return float(value)
    if isinstance(value, np.bool_):
        return bool(value)
    return value


def clean_document(obj):
    # Recursively clean dictionaries and lists before sending them to MongoDB.
    if isinstance(obj, dict):
        return {str(k): clean_document(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [clean_document(item) for item in obj]
    return clean_value(obj)


def row_to_dict(row):
    return clean_document(row.to_dict())


def stable_id(value):
    # Convert IDs into stable strings while handling missing values safely.
    if pd.isna(value):
        return None
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value)


def pick_existing(source_dict, fields):
    # Return only fields that exist and are not missing.
    result = {}
    for field in fields:
        if field in source_dict and source_dict[field] is not None:
            result[field] = source_dict[field]
    return result


print('Helper functions ready.')

Helper functions ready.


---
## **Prepare Master Collection Documents**

Each source row is converted into a MongoDB document. The business ID is used as `_id` so the notebook can be run again without creating duplicate records.

In [ ]:
def make_master_documents(df, id_column):
    docs = []
    for _, row in df.iterrows():
        doc = row_to_dict(row)
        doc['_id'] = stable_id(row[id_column])
        docs.append(doc)
    return docs

customer_docs  = make_master_documents(customers, 'customer_id')
driver_docs    = make_master_documents(drivers, 'driver_id')
vehicle_docs   = make_master_documents(vehicles, 'vehicle_id')
hub_docs       = make_master_documents(hubs, 'hub_id')
app_event_docs = make_master_documents(app_events, 'event_id')

master_summary = pd.DataFrame({
    'collection': ['customers', 'drivers', 'vehicles', 'hubs', 'app_events'],
    'prepared_documents': [len(customer_docs), len(driver_docs), len(vehicle_docs), len(hub_docs), len(app_event_docs)]
})

master_summary

,collection,prepared_documents
0,customers,650
1,drivers,170
2,vehicles,120
3,hubs,8
4,app_events,640


---
## **Build Nested Service Case Documents**

Each `service_cases` document starts from an order. If the order has a delivery, complaints, incidents, or app events, those related records are nested inside the same document. This creates an integrated view of each service case.

In [ ]:
customers_by_id = {stable_id(row['customer_id']): row_to_dict(row) for _, row in customers.iterrows()}
drivers_by_id   = {stable_id(row['driver_id']): row_to_dict(row) for _, row in drivers.iterrows()}
vehicles_by_id  = {stable_id(row['vehicle_id']): row_to_dict(row) for _, row in vehicles.iterrows()}
hubs_by_id      = {stable_id(row['hub_id']): row_to_dict(row) for _, row in hubs.iterrows()}

deliveries_by_order = {}
for _, row in deliveries.iterrows():
    deliveries_by_order[stable_id(row['order_id'])] = row_to_dict(row)

complaints_by_order = {}
for _, row in complaints.iterrows():
    order_id = stable_id(row['order_id'])
    complaints_by_order.setdefault(order_id, []).append(row_to_dict(row))

incidents_by_delivery = {}
for _, row in incidents.iterrows():
    delivery_id = stable_id(row['delivery_id'])
    incidents_by_delivery.setdefault(delivery_id, []).append(row_to_dict(row))

app_events_by_order = {}
if 'order_id' in app_events.columns:
    for _, row in app_events.dropna(subset=['order_id']).iterrows():
        order_id = stable_id(row['order_id'])
        app_events_by_order.setdefault(order_id, []).append(row_to_dict(row))

print('Lookup dictionaries created:')
print('deliveries linked to orders:', len(deliveries_by_order))
print('orders with complaints:', len(complaints_by_order))
print('deliveries with incidents:', len(incidents_by_delivery))
print('orders with app events:', len(app_events_by_order))

Lookup dictionaries created:
deliveries linked to orders: 950
orders with complaints: 285
deliveries with incidents: 248
orders with app events: 413


In [ ]:
service_case_docs = []

for _, order_row in orders.iterrows():
    order = row_to_dict(order_row)
    order_id = stable_id(order_row['order_id'])
    customer_id = stable_id(order_row['customer_id'])

    delivery = deliveries_by_order.get(order_id)
    delivery_id = None if delivery is None else stable_id(delivery.get('delivery_id'))

    related_complaints = complaints_by_order.get(order_id, [])

    related_incidents = []
    if delivery_id is not None:
        related_incidents = incidents_by_delivery.get(delivery_id, [])

    related_app_events = app_events_by_order.get(order_id, [])
    customer = customers_by_id.get(customer_id, {})

    driver_summary = None
    vehicle_summary = None
    hub_summary = None

    if delivery is not None:
        driver_summary = drivers_by_id.get(stable_id(delivery.get('driver_id')))
        vehicle_summary = vehicles_by_id.get(stable_id(delivery.get('vehicle_id')))
        hub_summary = hubs_by_id.get(stable_id(delivery.get('hub_id')))

    order_value = order.get('order_value')
    promised_window = order.get('promised_window_hours')

    delivery_status = None if delivery is None else delivery.get('delivery_status')
    completion_hours_clean = None if delivery is None else delivery.get('completion_hours_clean')
    cost_per_km = None if delivery is None else delivery.get('cost_per_km')
    customer_rating = None if delivery is None else delivery.get('customer_rating_post_delivery')
    manual_overrides = None if delivery is None else delivery.get('manual_route_override_count')
    is_bad_outcome = False if delivery is None else bool(delivery.get('is_bad_outcome') == 1 or delivery.get('is_bad_outcome') is True)
    missed_window = None if delivery is None else delivery.get('missed_promised_window')

    doc = {
        '_id': order_id,
        'order_id': order_id,
        'customer_id': customer_id,
        'service_type': order.get('service_type'),
        'priority_level': order.get('priority_level'),
        'zones': {
            'pickup': order.get('pickup_zone_clean'),
            'dropoff': order.get('dropoff_zone_clean')
        },
        'order': order,
        'customer_summary': pick_existing(customer, [
            'customer_id', 'customer_type', 'home_zone_clean',
            'loyalty_score', 'preferred_channel'
        ]),
        'delivery': delivery,
        'driver_summary': None if driver_summary is None else pick_existing(driver_summary, [
            'driver_id', 'employment_type', 'base_zone_clean',
            'years_experience', 'training_score', 'driver_rating'
        ]),
        'vehicle_summary': None if vehicle_summary is None else pick_existing(vehicle_summary, [
            'vehicle_id', 'vehicle_type', 'assigned_zone_clean',
            'battery_health_pct', 'maintenance_status', 'low_battery_flag'
        ]),
        'hub_summary': None if hub_summary is None else pick_existing(hub_summary, [
            'hub_id', 'hub_name', 'zone_clean', 'hub_type', 'capacity_score'
        ]),
        'complaints': related_complaints,
        'incidents': related_incidents,
        'app_events': related_app_events,
        'analytics': {
            'order_value': order_value,
            'promised_window_hours': promised_window,
            'delivery_status': delivery_status,
            'completion_hours_clean': completion_hours_clean,
            'cost_per_km': cost_per_km,
            'customer_rating_post_delivery': customer_rating,
            'manual_route_override_count': manual_overrides,
            'complaint_count': len(related_complaints),
            'incident_count': len(related_incidents),
            'app_event_count': len(related_app_events)
        },
        'status_flags': {
            'has_delivery': delivery is not None,
            'tracking_gap_flag': delivery is None,
            'has_complaint': len(related_complaints) > 0,
            'has_incident': len(related_incidents) > 0,
            'has_app_events': len(related_app_events) > 0,
            'is_bad_outcome': is_bad_outcome,
            'missed_promised_window': missed_window
        }
    }

    service_case_docs.append(clean_document(doc))

service_case_summary = pd.DataFrame({
    'collection': ['service_cases'],
    'prepared_documents': [len(service_case_docs)],
    'orders_without_delivery': [sum(1 for doc in service_case_docs if doc['status_flags']['tracking_gap_flag'])],
    'cases_with_complaints': [sum(1 for doc in service_case_docs if doc['status_flags']['has_complaint'])],
    'cases_with_incidents': [sum(1 for doc in service_case_docs if doc['status_flags']['has_incident'])]
})

service_case_summary

,collection,prepared_documents,orders_without_delivery,cases_with_complaints,cases_with_incidents
0,service_cases,1250,300,285,248


---
## **Connect to MongoDB Atlas**

The MongoDB Atlas connection string was written directly in the notebook code as required for the coursework demonstration. This allows the connection method to be clearly shown when the notebook is reviewed.

The connection string follows the MongoDB Atlas SRV format and connects Google Colab to the `NorthStarDB` database through PyMongo.

If the connection fails, the MongoDB Atlas Network Access settings should be checked. For Google Colab, access may need to allow `0.0.0.0/0`. The database username, password and cluster status should also be checked.

In [ ]:
# MongoDB Atlas connection string

MONGO_URI = "mongodb+srv://Samson5827:DBA_2026@cluster0.gbco7lz.mongodb.net/?appName=Cluster0"

client = MongoClient(
    MONGO_URI,
    tls=True,
    tlsCAFile=certifi.where(),
    serverSelectionTimeoutMS=30000
)

try:
    client.admin.command('ping')
    print('MongoDB Atlas connection successful.')
except ServerSelectionTimeoutError as error:
    print('Connection failed. Check IP access list, username, password, and cluster status.')
    raise error
except OperationFailure as error:
    print('Authentication failed. Check username and password.')
    raise error

DB_NAME = 'NorthStarDB'
db = client[DB_NAME]

print('Using database:', DB_NAME)

MongoDB Atlas connection successful.
Using database: NorthStarDB


---
## **Insert or Update Documents in MongoDB Atlas**

The insertion function uses `ReplaceOne` with `upsert=True`. This means the notebook can be run more than once without creating duplicate documents. If a document with the same `_id` already exists, it is replaced with the latest version.

In [ ]:
def upsert_documents(collection_name, docs, batch_size=500):
    collection = db[collection_name]
    total_matched = 0
    total_modified = 0
    total_upserted = 0

    for start in range(0, len(docs), batch_size):
        batch = docs[start:start + batch_size]
        operations = [ReplaceOne({'_id': doc['_id']}, doc, upsert=True) for doc in batch]

        if operations:
            try:
                result = collection.bulk_write(operations, ordered=False)
                total_matched += result.matched_count
                total_modified += result.modified_count
                total_upserted += len(result.upserted_ids)
            except BulkWriteError as error:
                print('Bulk write error in collection:', collection_name)
                raise error

    return {
        'collection': collection_name,
        'documents_processed': len(docs),
        'documents_matched': total_matched,
        'documents_modified': total_modified,
        'documents_upserted': total_upserted,
        'final_count': collection.count_documents({})
    }

insert_results = [
    upsert_documents('customers', customer_docs),
    upsert_documents('drivers', driver_docs),
    upsert_documents('vehicles', vehicle_docs),
    upsert_documents('hubs', hub_docs),
    upsert_documents('app_events', app_event_docs),
    upsert_documents('service_cases', service_case_docs)
]

insert_summary = pd.DataFrame(insert_results)
insert_summary

,collection,documents_processed,documents_matched,documents_modified,documents_upserted,final_count
0,customers,650,650,650,0,650
1,drivers,170,170,170,0,170
2,vehicles,120,120,120,0,120
3,hubs,8,8,8,0,8
4,app_events,640,640,640,0,640
5,service_cases,1250,1250,0,0,1250


---
## **Confirm Collections and View One Service Case**

This check confirms that the MongoDB collections were created and populated. The sample document is useful evidence because it shows the nested NoSQL structure used for the NorthStar case study.

In [ ]:
collections = sorted(db.list_collection_names())

print('Collections in NorthStarDB:')
for collection_name in collections:
    print(f'  {collection_name}: {db[collection_name].count_documents({})} documents')

print('\nSample service case document:')
sample_case = db.service_cases.find_one(
    {'status_flags.has_complaint': True, 'status_flags.has_delivery': True},
    {
        '_id': 1,
        'order_id': 1,
        'customer_id': 1,
        'service_type': 1,
        'zones': 1,
        'delivery.delivery_id': 1,
        'delivery.delivery_status': 1,
        'delivery.completion_hours_clean': 1,
        'complaints.complaint_type': 1,
        'incidents.incident_type': 1,
        'app_events.event_type': 1,
        'status_flags': 1,
        'analytics': 1
    }
)

pprint(sample_case)

Collections in NorthStarDB:
  app_events: 640 documents
  customers: 650 documents
  drivers: 170 documents
  hubs: 8 documents
  service_cases: 1250 documents
  vehicles: 120 documents

Sample service case document:
{'_id': 'O00003',
 'analytics': {'app_event_count': 0,
               'complaint_count': 1,
               'completion_hours_clean': 8.861012409166666,
               'cost_per_km': 1.0092024539877302,
               'customer_rating_post_delivery': 3.7,
               'delivery_status': 'Delayed',
               'incident_count': 0,
               'manual_route_override_count': 2,
               'order_value': 33.5,
               'promised_window_hours': 4},
 'app_events': [],
 'complaints': [{'complaint_type': 'AppIssue'}],
 'customer_id': 'C0161',
 'delivery': {'completion_hours_clean': 8.861012409166666,
              'delivery_id': 'DL00925',
              'delivery_status': 'Delayed'},
 'incidents': [],
 'order_id': 'O00003',
 'service_type': 'Passenger',
 'status_f

---
## **Safe CRUD Demonstration**

The coursework requires MongoDB CRUD operations. To avoid changing real NorthStar records, this demonstration creates a temporary copy of an existing `service_cases` document. The copied document is inserted with a new temporary ID, read back, updated, deleted, and then checked again to confirm that it has been removed.

This approach keeps the CRUD demonstration connected to the real NorthStar dataset while protecting the original service case records.

In [ ]:
# Delete any old temporary CRUD copy first so this cell can run more than once.
TEMP_CRUD_ID = 'CRUD_COPY_SERVICE_CASE'
db.service_cases.delete_one({'_id': TEMP_CRUD_ID})

# Select one existing real service case to copy.
source_case = db.service_cases.find_one({
    'status_flags.has_delivery': True,
    'status_flags.has_complaint': True
})

if source_case is None:
    source_case = db.service_cases.find_one({'_id': {'$ne': TEMP_CRUD_ID}})

if source_case is None:
    raise ValueError('No source service case was found for the CRUD copy demonstration.')

# Create a temporary copy of the real service case.
# The original document is not changed.
crud_case = copy.deepcopy(source_case)
source_case_id = source_case.get('_id')
source_order_id = source_case.get('order_id')

crud_case['_id'] = TEMP_CRUD_ID
crud_case['order_id'] = TEMP_CRUD_ID
crud_case['source_case_id'] = source_case_id
crud_case['source_order_id'] = source_order_id
crud_case['case_review_status'] = 'New'
crud_case['review_note'] = 'Temporary CRUD copy created from an existing NorthStar service case.'

if isinstance(crud_case.get('order'), dict):
    crud_case['order']['order_id'] = TEMP_CRUD_ID

def crud_summary(document):
    if document is None:
        return None

    return {
        '_id': document.get('_id'),
        'order_id': document.get('order_id'),
        'source_case_id': document.get('source_case_id'),
        'source_order_id': document.get('source_order_id'),
        'service_type': document.get('service_type'),
        'priority_level': document.get('priority_level'),
        'zones': document.get('zones'),
        'delivery_status': (document.get('delivery') or {}).get('delivery_status'),
        'complaint_count': len(document.get('complaints', [])),
        'incident_count': len(document.get('incidents', [])),
        'case_review_status': document.get('case_review_status'),
        'review_note': document.get('review_note')
    }

print('Source real service case selected for temporary CRUD copy:')
pprint(crud_summary(source_case))

# INSERT
insert_result = db.service_cases.insert_one(crud_case)
print('\nINSERT result inserted_id:', insert_result.inserted_id)

# READ
print('\nREAD result for temporary copied service case:')
read_result = db.service_cases.find_one({'_id': TEMP_CRUD_ID})
pprint(crud_summary(read_result))

# UPDATE
update_result = db.service_cases.update_one(
    {'_id': TEMP_CRUD_ID},
    {
        '$set': {
            'case_review_status': 'Reviewed',
            'review_note': 'Temporary copied service case updated successfully during CRUD demonstration.'
        }
    }
)
print('\nUPDATE result modified_count:', update_result.modified_count)

print('\nREAD after update:')
updated_result = db.service_cases.find_one({'_id': TEMP_CRUD_ID})
pprint(crud_summary(updated_result))

# DELETE
delete_result = db.service_cases.delete_one({'_id': TEMP_CRUD_ID})
print('\nDELETE result deleted_count:', delete_result.deleted_count)

print('\nREAD after delete:')
pprint(db.service_cases.find_one({'_id': TEMP_CRUD_ID}))

print('\nThe CRUD operation used a temporary copy of a real NorthStar service case. The original source service case was not modified.')


Source real service case selected for temporary CRUD copy:
{'_id': 'O00003',
 'case_review_status': None,
 'complaint_count': 1,
 'delivery_status': 'Delayed',
 'incident_count': 0,
 'order_id': 'O00003',
 'priority_level': 'High',
 'review_note': None,
 'service_type': 'Passenger',
 'source_case_id': None,
 'source_order_id': None,
 'zones': {'dropoff': 'Airport', 'pickup': 'West'}}

INSERT result inserted_id: CRUD_COPY_SERVICE_CASE

READ result for temporary copied service case:
{'_id': 'CRUD_COPY_SERVICE_CASE',
 'case_review_status': 'New',
 'complaint_count': 1,
 'delivery_status': 'Delayed',
 'incident_count': 0,
 'order_id': 'CRUD_COPY_SERVICE_CASE',
 'priority_level': 'High',
 'review_note': 'Temporary CRUD copy created from an existing NorthStar '
                'service case.',
 'service_type': 'Passenger',
 'source_case_id': 'O00003',
 'source_order_id': 'O00003',
 'zones': {'dropoff': 'Airport', 'pickup': 'West'}}

UPDATE result modified_count: 1

READ after update:
{'_id':

---
## **MongoDB Find Queries for Business Questions**

These queries show how the document model supports direct operational investigation. The same `service_cases` collection can answer questions about zones, delivery outcomes, complaints, incidents, and tracking gaps.

In [ ]:
central_bad_cases = list(db.service_cases.find(
    {'zones.pickup': 'Central', 'status_flags.is_bad_outcome': True},
    {
        '_id': 1,
        'service_type': 1,
        'priority_level': 1,
        'zones': 1,
        'delivery.delivery_status': 1,
        'delivery.completion_hours_clean': 1,
        'delivery.customer_rating_post_delivery': 1,
        'analytics.complaint_count': 1
    }
).limit(10))

print('Central bad outcome cases shown:', len(central_bad_cases))
pprint(central_bad_cases[:3])

Central bad outcome cases shown: 10
[{'_id': 'O00007',
  'analytics': {'complaint_count': 1},
  'delivery': {'completion_hours_clean': 8.921543125,
               'customer_rating_post_delivery': 3.93,
               'delivery_status': 'Delayed'},
  'priority_level': 'Low',
  'service_type': 'Business',
  'zones': {'dropoff': 'Airport', 'pickup': 'Central'}},
 {'_id': 'O00059',
  'analytics': {'complaint_count': 0},
  'delivery': {'completion_hours_clean': None,
               'customer_rating_post_delivery': 3.32,
               'delivery_status': 'Delayed'},
  'priority_level': 'Medium',
  'service_type': 'Passenger',
  'zones': {'dropoff': 'Airport', 'pickup': 'Central'}},
 {'_id': 'O00070',
  'analytics': {'complaint_count': 0},
  'delivery': {'completion_hours_clean': 16.55390397277778,
               'customer_rating_post_delivery': 2.14,
               'delivery_status': 'Failed'},
  'priority_level': 'Low',
  'service_type': 'Retail',
  'zones': {'dropoff': 'Airport', 'pickup':

In [ ]:
tracking_gap_count = db.service_cases.count_documents({'status_flags.tracking_gap_flag': True})

tracking_gap_cases = list(db.service_cases.find(
    {'status_flags.tracking_gap_flag': True},
    {
        '_id': 1,
        'customer_id': 1,
        'service_type': 1,
        'priority_level': 1,
        'zones': 1,
        'order.order_value': 1,
        'analytics.complaint_count': 1
    }
).limit(10))

print('Total tracking gap cases:', tracking_gap_count)
pprint(tracking_gap_cases[:3])

Total tracking gap cases: 300
[{'_id': 'O00002',
  'analytics': {'complaint_count': 0},
  'customer_id': 'C0459',
  'order': {'order_value': 109.3},
  'priority_level': 'Low',
  'service_type': 'Passenger',
  'zones': {'dropoff': 'Airport', 'pickup': 'North'}},
 {'_id': 'O00006',
  'analytics': {'complaint_count': 0},
  'customer_id': 'C0437',
  'order': {'order_value': 151.44},
  'priority_level': 'High',
  'service_type': 'Retail',
  'zones': {'dropoff': 'East', 'pickup': 'Central'}},
 {'_id': 'O00011',
  'analytics': {'complaint_count': 1},
  'customer_id': 'C0340',
  'order': {'order_value': 79.1},
  'priority_level': 'Low',
  'service_type': 'Parcel',
  'zones': {'dropoff': 'Central', 'pickup': 'West'}}]


In [ ]:
bad_with_complaints_count = db.service_cases.count_documents({
    'status_flags.is_bad_outcome': True,
    'status_flags.has_complaint': True
})

bad_with_complaints = list(db.service_cases.find(
    {'status_flags.is_bad_outcome': True, 'status_flags.has_complaint': True},
    {
        '_id': 1,
        'service_type': 1,
        'zones': 1,
        'delivery.delivery_status': 1,
        'complaints.complaint_type': 1,
        'complaints.severity': 1,
        'analytics.complaint_count': 1
    }
).limit(10))

print('Bad outcome cases with complaints:', bad_with_complaints_count)
pprint(bad_with_complaints[:3])

Bad outcome cases with complaints: 78
[{'_id': 'O00003',
  'analytics': {'complaint_count': 1},
  'complaints': [{'complaint_type': 'AppIssue', 'severity': 'Medium'}],
  'delivery': {'delivery_status': 'Delayed'},
  'service_type': 'Passenger',
  'zones': {'dropoff': 'Airport', 'pickup': 'West'}},
 {'_id': 'O00007',
  'analytics': {'complaint_count': 1},
  'complaints': [{'complaint_type': 'AppIssue', 'severity': 'High'}],
  'delivery': {'delivery_status': 'Delayed'},
  'service_type': 'Business',
  'zones': {'dropoff': 'Airport', 'pickup': 'Central'}},
 {'_id': 'O00023',
  'analytics': {'complaint_count': 1},
  'complaints': [{'complaint_type': 'SupportExperience', 'severity': 'Low'}],
  'delivery': {'delivery_status': 'Failed'},
  'service_type': 'Passenger',
  'zones': {'dropoff': 'North', 'pickup': 'North'}}]


---
## **MongoDB Aggregation Pipelines**

Aggregation pipelines summarise the MongoDB documents in a way that supports report findings. These examples use `$match`, `$group`, `$unwind`, `$avg`, `$sum`, `$addFields`, and `$sort`.

In [ ]:
pipeline_zone_bad_outcomes = [
    {'$match': {'status_flags.has_delivery': True}},
    {'$group': {
        '_id': '$zones.pickup',
        'delivery_records': {'$sum': 1},
        'bad_outcomes': {'$sum': {'$cond': ['$status_flags.is_bad_outcome', 1, 0]}},
        'complaint_cases': {'$sum': {'$cond': ['$status_flags.has_complaint', 1, 0]}},
        'avg_rating': {'$avg': '$delivery.customer_rating_post_delivery'},
        'avg_completion_hours': {'$avg': '$delivery.completion_hours_clean'}
    }},
    {'$addFields': {
        'bad_outcome_pct': {'$round': [{'$multiply': [{'$divide': ['$bad_outcomes', '$delivery_records']}, 100]}, 1]},
        'complaint_pct': {'$round': [{'$multiply': [{'$divide': ['$complaint_cases', '$delivery_records']}, 100]}, 1]}
    }},
    {'$sort': {'bad_outcome_pct': -1}}
]

agg_zone_bad = list(db.service_cases.aggregate(pipeline_zone_bad_outcomes))
agg_zone_bad_df = pd.DataFrame(agg_zone_bad).rename(columns={'_id': 'pickup_zone'})
agg_zone_bad_df

,pickup_zone,delivery_records,bad_outcomes,complaint_cases,avg_rating,avg_completion_hours,bad_outcome_pct,complaint_pct
0,Central,174,84,37,3.546036,11.245158,48.3,21.3
1,Airport,113,43,23,3.984037,9.648656,38.1,20.4
2,Riverside,119,43,31,3.864492,10.756360,36.1,26.1
3,East,156,50,35,3.912078,10.428183,32.1,22.4
4,North,135,43,34,3.896667,9.785670,31.9,25.2
5,West,114,35,17,3.896316,10.160827,30.7,14.9
6,South,139,36,32,4.051825,9.848314,25.9,23.0


In [ ]:
pipeline_complaint_types = [
    {'$unwind': '$complaints'},
    {'$group': {
        '_id': '$complaints.complaint_type',
        'complaint_count': {'$sum': 1},
        'avg_resolution_days': {'$avg': '$complaints.resolution_days'},
        'total_compensation': {'$sum': '$complaints.compensation_amount'}
    }},
    {'$sort': {'complaint_count': -1}}
]

agg_complaints = list(db.service_cases.aggregate(pipeline_complaint_types))
agg_complaints_df = pd.DataFrame(agg_complaints).rename(columns={'_id': 'complaint_type'})
agg_complaints_df

,complaint_type,complaint_count,avg_resolution_days,total_compensation
0,Delay,101,7.257426,1696.84
1,MissedPickup,64,7.640625,1423.40
2,AppIssue,53,8.603774,980.72
3,DriverBehaviour,51,8.156863,973.06
4,SupportExperience,20,7.450000,342.50
5,Billing,16,7.750000,381.94
6,Damage,15,11.333333,359.73


In [ ]:
pipeline_hub_performance = [
    {'$match': {'status_flags.has_delivery': True}},
    {'$group': {
        '_id': '$hub_summary.hub_name',
        'hub_zone': {'$first': '$hub_summary.zone_clean'},
        'delivery_records': {'$sum': 1},
        'bad_outcomes': {'$sum': {'$cond': ['$status_flags.is_bad_outcome', 1, 0]}},
        'avg_cost_per_km': {'$avg': '$delivery.cost_per_km'},
        'avg_rating': {'$avg': '$delivery.customer_rating_post_delivery'},
        'avg_manual_overrides': {'$avg': '$delivery.manual_route_override_count'}
    }},
    {'$addFields': {
        'bad_outcome_pct': {'$round': [{'$multiply': [{'$divide': ['$bad_outcomes', '$delivery_records']}, 100]}, 1]}
    }},
    {'$sort': {'bad_outcome_pct': -1}}
]

agg_hub = list(db.service_cases.aggregate(pipeline_hub_performance))
agg_hub_df = pd.DataFrame(agg_hub).rename(columns={'_id': 'hub_name'})
agg_hub_df

,hub_name,hub_zone,delivery_records,bad_outcomes,avg_cost_per_km,avg_rating,avg_manual_overrides,bad_outcome_pct
0,Central Core,Central,115,48,1.477564,3.669558,0.947826,41.7
1,Airport Hub,Airport,104,42,1.239235,3.882136,0.913462,40.4
2,Midtown Relay,Central,128,48,1.181762,3.884560,1.109375,37.5
3,West Gate,West,127,44,1.355406,3.915476,0.874016,34.6
4,South Link,South,106,36,1.168068,3.950952,0.915094,34.0
5,Riverside Hub,Riverside,115,39,1.222085,3.881858,1.052174,33.9
6,North Exchange,North,136,43,1.371519,3.840593,1.029412,31.6
7,East Dock,East,119,34,1.054579,3.895862,0.890756,28.6


In [ ]:
pipeline_app_latency = [
    {'$group': {
        '_id': '$event_type',
        'event_count': {'$sum': 1},
        'avg_latency_ms': {'$avg': '$api_latency_ms'},
        'success_rate': {'$avg': '$success_flag'}
    }},
    {'$addFields': {
        'success_rate_pct': {'$round': [{'$multiply': ['$success_rate', 100]}, 1]},
        'avg_latency_ms': {'$round': ['$avg_latency_ms', 1]}
    }},
    {'$sort': {'avg_latency_ms': -1}}
]

agg_app_latency = list(db.app_events.aggregate(pipeline_app_latency))
agg_app_latency_df = pd.DataFrame(agg_app_latency).rename(columns={'_id': 'event_type'})
agg_app_latency_df

,event_type,event_count,avg_latency_ms,success_rate,success_rate_pct
0,delivery_instruction_update,75,496.3,1.000000,100.0
1,chat_opened,88,478.3,1.000000,100.0
2,chat_escalated,38,478.1,0.500000,50.0
3,payment_retry,69,472.7,0.724638,72.5
4,track_order,138,460.7,1.000000,100.0
5,search_route,99,456.5,1.000000,100.0
6,eta_refresh,105,452.2,1.000000,100.0
7,cancel_attempt,28,417.1,1.000000,100.0


---
## **Save MongoDB Outputs Locally for Coursework Evidence**

The key MongoDB query and aggregation results are saved locally in the Colab session. A zip file is created to preserve the generated MongoDB output evidence under `Outputs/MongoDB_Outputs`.

In [ ]:
MONGO_OUTPUT_PATH = Path('/content/northstar_mongodb_outputs')
MONGO_OUTPUT_PATH.mkdir(parents=True, exist_ok=True)

collection_counts = pd.DataFrame([
    {'collection': name, 'document_count': db[name].count_documents({})}
    for name in sorted(db.list_collection_names())
])

central_bad_df = pd.DataFrame(central_bad_cases)
tracking_gap_df = pd.DataFrame(tracking_gap_cases)
bad_with_complaints_df = pd.DataFrame(bad_with_complaints)

collection_counts.to_csv(MONGO_OUTPUT_PATH / 'mongodb_collection_counts.csv', index=False)
insert_summary.to_csv(MONGO_OUTPUT_PATH / 'mongodb_insert_summary.csv', index=False)
master_summary.to_csv(MONGO_OUTPUT_PATH / 'mongodb_master_preparation_summary.csv', index=False)
service_case_summary.to_csv(MONGO_OUTPUT_PATH / 'mongodb_service_case_preparation_summary.csv', index=False)
central_bad_df.to_csv(MONGO_OUTPUT_PATH / 'mongodb_find_central_bad_cases.csv', index=False)
tracking_gap_df.to_csv(MONGO_OUTPUT_PATH / 'mongodb_find_tracking_gap_cases.csv', index=False)
bad_with_complaints_df.to_csv(MONGO_OUTPUT_PATH / 'mongodb_find_bad_with_complaints.csv', index=False)
agg_zone_bad_df.to_csv(MONGO_OUTPUT_PATH / 'mongodb_bad_outcomes_by_zone.csv', index=False)
agg_complaints_df.to_csv(MONGO_OUTPUT_PATH / 'mongodb_complaints_by_type.csv', index=False)
agg_hub_df.to_csv(MONGO_OUTPUT_PATH / 'mongodb_hub_performance.csv', index=False)
agg_app_latency_df.to_csv(MONGO_OUTPUT_PATH / 'mongodb_app_event_latency.csv', index=False)

with open(MONGO_OUTPUT_PATH / 'mongodb_sample_service_case.json', 'w', encoding='utf-8') as file:
    file.write(json_util.dumps(sample_case, indent=2))

print('MongoDB outputs saved locally in Colab:')
print(MONGO_OUTPUT_PATH)
print()

for file_path in sorted(MONGO_OUTPUT_PATH.glob('*')):
    if file_path.is_file():
        size_kb = file_path.stat().st_size // 1024
        print(f'{file_path.name:<45} {size_kb} KB')

zip_base = '/content/northstar_mongodb_outputs'
zip_file = zip_base + '.zip'

if Path(zip_file).exists():
    Path(zip_file).unlink()

zip_file = shutil.make_archive(zip_base, 'zip', MONGO_OUTPUT_PATH)

print('\nMongoDB output zip created:')
print(zip_file)

MongoDB outputs saved locally in Colab:
/content/northstar_mongodb_outputs

mongodb_app_event_latency.csv                 0 KB
mongodb_bad_outcomes_by_zone.csv              0 KB
mongodb_collection_counts.csv                 0 KB
mongodb_complaints_by_type.csv                0 KB
mongodb_find_bad_with_complaints.csv          1 KB
mongodb_find_central_bad_cases.csv            2 KB
mongodb_find_tracking_gap_cases.csv           1 KB
mongodb_hub_performance.csv                   0 KB
mongodb_insert_summary.csv                    0 KB
mongodb_master_preparation_summary.csv        0 KB
mongodb_sample_service_case.json              0 KB
mongodb_service_case_preparation_summary.csv  0 KB

MongoDB output zip created:
/content/northstar_mongodb_outputs.zip


# **Section 2 - Query Optimisation/Data Manipulation in MongoDB**

This section evaluates important MongoDB queries using explain plans, creates targeted indexes, and compares execution statistics before and after indexing. The aim is to show how query optimisation can support faster operational reporting for NorthStar.

## **Select Collections and Check Document Counts**

Before optimisation, the notebook checks that the collections created in the MongoDB Design section above are available and contain the expected number of documents.

In [ ]:
service_cases = db['service_cases']
app_events = db['app_events']

collections_to_check = ['service_cases', 'app_events', 'customers', 'drivers', 'vehicles', 'hubs']

collection_counts = []
for name in collections_to_check:
    count = db[name].count_documents({})
    collection_counts.append({'collection': name, 'document_count': count})

collection_counts_df = pd.DataFrame(collection_counts)
collection_counts_df

,collection,document_count
0,service_cases,1250
1,app_events,640
2,customers,650
3,drivers,170
4,vehicles,120
5,hubs,8


## **Define Helper Functions for Explain Plans**

MongoDB explain plans can contain nested structures. These helper functions extract the most useful performance indicators:

- `nReturned`
- `totalDocsExamined`
- `totalKeysExamined`
- `executionTimeMillis`
- plan stages such as `COLLSCAN` and `IXSCAN`

A collection scan means MongoDB checked documents without using a suitable index. An index scan means MongoDB used an index to locate matching documents more efficiently.

In [ ]:
def collect_values(obj, target_key):
    values = []

    if isinstance(obj, dict):
        for key, value in obj.items():
            if key == target_key and isinstance(value, (int, float)):
                values.append(value)
            values.extend(collect_values(value, target_key))
    elif isinstance(obj, list):
        for item in obj:
            values.extend(collect_values(item, target_key))

    return values


def collect_plan_stages(obj):
    stages = []

    if isinstance(obj, dict):
        if 'stage' in obj:
            stages.append(str(obj['stage']))
        for value in obj.values():
            stages.extend(collect_plan_stages(value))
    elif isinstance(obj, list):
        for item in obj:
            stages.extend(collect_plan_stages(item))

    return stages


def safe_query_name(name):
    return re.sub(r'[^A-Za-z0-9_]+', '_', name).strip('_').lower()


def summarise_explain(query_name, collection_name, explain_doc, index_status):
    execution_stats = explain_doc.get('executionStats', {})
    query_planner = explain_doc.get('queryPlanner', {})

    # Normal find explain plans usually keep execution statistics at the top level.
    # Aggregation explain plans can place the same details deeper inside the document.
    # These fallback sources allow the same helper to work for both query types.
    stats_source = execution_stats if execution_stats else explain_doc
    planner_source = query_planner if query_planner else explain_doc

    n_returned_values = collect_values(stats_source, 'nReturned')
    docs_examined_values = collect_values(stats_source, 'totalDocsExamined')
    keys_examined_values = collect_values(stats_source, 'totalKeysExamined')
    time_values = collect_values(stats_source, 'executionTimeMillis')

    stages = collect_plan_stages(planner_source)
    unique_stages = sorted(set(stages))

    uses_index = any('IXSCAN' in stage or 'EXPRESS_IXSCAN' in stage for stage in unique_stages)
    uses_collection_scan = any('COLLSCAN' in stage for stage in unique_stages)

    return {
        'query_name': query_name,
        'collection': collection_name,
        'index_status': index_status,
        'n_returned': max(n_returned_values) if n_returned_values else None,
        'total_docs_examined': max(docs_examined_values) if docs_examined_values else None,
        'total_keys_examined': max(keys_examined_values) if keys_examined_values else None,
        'execution_time_ms': max(time_values) if time_values else None,
        'uses_index': uses_index,
        'uses_collection_scan': uses_collection_scan,
        'plan_stages': ', '.join(unique_stages)
    }


def run_find_explain(collection, query, projection=None, sort=None, limit=25):
    command = {
        'find': collection.name,
        'filter': query,
        'projection': projection or {},
        'limit': limit
    }

    if sort is not None:
        command['sort'] = dict(sort)

    explain_doc = db.command('explain', command, verbosity='executionStats')
    return explain_doc


def save_json(path, obj):
    with open(path, 'w', encoding='utf-8') as file:
        json.dump(obj, file, indent=2, default=str)


def flatten_dict(data, parent_key='', sep='.'):
    items = {}

    for key, value in data.items():
        new_key = f'{parent_key}{sep}{key}' if parent_key else key

        if isinstance(value, dict):
            items.update(flatten_dict(value, new_key, sep=sep))
        elif isinstance(value, list):
            items[new_key] = json.dumps(value, default=str)
        else:
            items[new_key] = value

    return items


def pct_reduction(before_value, after_value):
    if pd.isna(before_value) or pd.isna(after_value) or before_value == 0:
        return None
    return round(((before_value - after_value) / before_value) * 100, 2)


print('Helper functions ready.')

Helper functions ready.


## **Define Business Queries for Optimisation**

The chosen queries reflect the main business problems from the NorthStar case study:

1. High risk bad outcomes in the Central pickup zone
2. Complaint cases linked to delay complaints
3. Orders with tracking gaps caused by missing delivery records
4. High latency app events for delivery instruction updates

These queries are suitable for optimisation because they are realistic searches that managers may use repeatedly.

In [ ]:
query_specs = [
    {
        'query_name': 'Central bad outcome cases',
        'collection': service_cases,
        'filter': {
            'zones.pickup': 'Central',
            'status_flags.is_bad_outcome': True
        },
        'projection': {
            '_id': 1,
            'order_id': 1,
            'zones.pickup': 1,
            'analytics.completion_hours_clean': 1,
            'analytics.customer_rating_post_delivery': 1,
            'status_flags.is_bad_outcome': 1
        },
        'sort': [('analytics.completion_hours_clean', DESCENDING)],
        'limit': 25
    },
    {
        'query_name': 'Delay complaint service cases',
        'collection': service_cases,
        'filter': {
            'status_flags.has_complaint': True,
            'complaints.complaint_type': 'Delay'
        },
        'projection': {
            '_id': 1,
            'order_id': 1,
            'service_type': 1,
            'complaints.complaint_type': 1,
            'complaints.severity': 1,
            'analytics.customer_rating_post_delivery': 1
        },
        'sort': None,
        'limit': 25
    },
    {
        'query_name': 'Tracking gap cases',
        'collection': service_cases,
        'filter': {
            'status_flags.tracking_gap_flag': True
        },
        'projection': {
            '_id': 1,
            'order_id': 1,
            'service_type': 1,
            'zones.pickup': 1,
            'status_flags.tracking_gap_flag': 1
        },
        'sort': None,
        'limit': 25
    },
    {
        'query_name': 'Delivery instruction update latency',
        'collection': app_events,
        'filter': {
            'event_type': 'delivery_instruction_update'
        },
        'projection': {
            '_id': 1,
            'event_id': 1,
            'order_id': 1,
            'event_type': 1,
            'api_latency_ms': 1,
            'success_flag': 1
        },
        'sort': [('api_latency_ms', DESCENDING)],
        'limit': 25
    }
]

print('Business queries prepared for optimisation:')
for spec in query_specs:
    print('-', spec['query_name'])

Business queries prepared for optimisation:
- Central bad outcome cases
- Delay complaint service cases
- Tracking gap cases
- Delivery instruction update latency


## **Remove Existing Non-Default Indexes for a Fair Before Test**

To make the before and after comparison clear, this step removes previous custom indexes from the two target collections. The default `_id` index is kept because MongoDB requires it.

In [ ]:
def drop_custom_indexes(collection):
    existing_indexes = list(collection.list_indexes())
    dropped = []

    for index in existing_indexes:
        index_name = index.get('name')
        if index_name != '_id_':
            collection.drop_index(index_name)
            dropped.append(index_name)

    return dropped


dropped_service_indexes = drop_custom_indexes(service_cases)
dropped_app_event_indexes = drop_custom_indexes(app_events)

print('Dropped custom indexes from service_cases:', dropped_service_indexes)
print('Dropped custom indexes from app_events:', dropped_app_event_indexes)

print('\nRemaining service_cases indexes:')
for index in service_cases.list_indexes():
    print(index)

print('\nRemaining app_events indexes:')
for index in app_events.list_indexes():
    print(index)

Dropped custom indexes from service_cases: []
Dropped custom indexes from app_events: []

Remaining service_cases indexes:
SON([('v', 2), ('key', SON([('_id', 1)])), ('name', '_id_')])

Remaining app_events indexes:
SON([('v', 2), ('key', SON([('_id', 1)])), ('name', '_id_')])


## **Run Explain Plans Before Indexing**

This step records the explain plan for each query before any custom index is created. The expected behaviour is that some queries may use collection scans because only the default `_id` index is available.

In [ ]:
OUTPUT_PATH = Path('/content/northstar_query_optimisation_outputs')
OUTPUT_PATH.mkdir(parents=True, exist_ok=True)

before_summaries = []

for spec in query_specs:
    explain_doc = run_find_explain(
        collection=spec['collection'],
        query=spec['filter'],
        projection=spec['projection'],
        sort=spec['sort'],
        limit=spec['limit']
    )

    summary = summarise_explain(
        query_name=spec['query_name'],
        collection_name=spec['collection'].name,
        explain_doc=explain_doc,
        index_status='before_index'
    )

    before_summaries.append(summary)

    json_path = OUTPUT_PATH / f"explain_before_{safe_query_name(spec['query_name'])}.json"
    save_json(json_path, explain_doc)

before_df = pd.DataFrame(before_summaries)
before_df

,query_name,collection,index_status,n_returned,total_docs_examined,total_keys_examined,execution_time_ms,uses_index,uses_collection_scan,plan_stages
0,Central bad outcome cases,service_cases,before_index,84,1250,0,2,False,True,"COLLSCAN, PROJECTION_DEFAULT, SORT"
1,Delay complaint service cases,service_cases,before_index,25,353,0,0,False,True,"COLLSCAN, LIMIT, PROJECTION_DEFAULT"
2,Tracking gap cases,service_cases,before_index,25,124,0,0,False,True,"COLLSCAN, LIMIT, PROJECTION_DEFAULT"
3,Delivery instruction update latency,app_events,before_index,75,640,0,0,False,True,"COLLSCAN, PROJECTION_SIMPLE, SORT"


## **Aggregation Pipeline Explain Before Indexing**

This step adds one MongoDB aggregation pipeline to the optimisation evidence. The pipeline filters Central zone bad outcome service cases and then groups the matching records by pickup zone.

This is useful because MongoDB coursework should not only show simple `find()` queries. It should also show how an aggregation pipeline can be examined and optimised. The `$match` stage is placed before `$group` so MongoDB can reduce the number of documents as early as possible.

In [ ]:
aggregation_pipeline = [
    {
        '$match': {
            'zones.pickup': 'Central',
            'status_flags.is_bad_outcome': True
        }
    },
    {
        '$group': {
            '_id': '$zones.pickup',
            'matching_bad_outcome_cases': {'$sum': 1},
            'avg_completion_hours': {'$avg': '$analytics.completion_hours_clean'},
            'avg_rating': {'$avg': '$analytics.customer_rating_post_delivery'}
        }
    },
    {
        '$sort': {
            'matching_bad_outcome_cases': -1
        }
    }
]

aggregation_explain_before = db.command(
    'explain',
    {
        'aggregate': 'service_cases',
        'pipeline': aggregation_pipeline,
        'cursor': {}
    },
    verbosity='executionStats'
)

aggregation_before_summary = summarise_explain(
    query_name='Aggregation zone bad outcome pipeline',
    collection_name='service_cases',
    explain_doc=aggregation_explain_before,
    index_status='before_index'
)

save_json(OUTPUT_PATH / 'aggregation_explain_before.json', aggregation_explain_before)

aggregation_before_df = pd.DataFrame([aggregation_before_summary])
aggregation_before_df


,query_name,collection,index_status,n_returned,total_docs_examined,total_keys_examined,execution_time_ms,uses_index,uses_collection_scan,plan_stages
0,Aggregation zone bad outcome pipeline,service_cases,before_index,1250,1250,0,2,False,True,"COLLSCAN, GROUP, filter, group, project, scan"


## **Create Indexes for the Target Queries**

Each index is selected based on the fields used in the query filters and sort conditions.

The aim is not to index every field. The aim is to create targeted indexes that support the most important repeated business queries.

In [ ]:
created_indexes = []

created_indexes.append({
    'collection': 'service_cases',
    'index_name': service_cases.create_index(
        [
            ('zones.pickup', ASCENDING),
            ('status_flags.is_bad_outcome', ASCENDING),
            ('analytics.completion_hours_clean', DESCENDING)
        ],
        name='idx_zone_bad_completion'
    ),
    'purpose': 'Supports filtering by pickup zone and bad outcome, with sorting by completion time.'
})

created_indexes.append({
    'collection': 'service_cases',
    'index_name': service_cases.create_index(
        [
            ('status_flags.has_complaint', ASCENDING),
            ('complaints.complaint_type', ASCENDING)
        ],
        name='idx_complaint_type'
    ),
    'purpose': 'Supports searches for embedded complaint types inside service cases.'
})

created_indexes.append({
    'collection': 'service_cases',
    'index_name': service_cases.create_index(
        [
            ('status_flags.tracking_gap_flag', ASCENDING)
        ],
        name='idx_tracking_gap'
    ),
    'purpose': 'Supports fast retrieval of orders that have no linked delivery record.'
})

created_indexes.append({
    'collection': 'app_events',
    'index_name': app_events.create_index(
        [
            ('event_type', ASCENDING),
            ('api_latency_ms', DESCENDING)
        ],
        name='idx_app_event_latency'
    ),
    'purpose': 'Supports filtering app events by event type and sorting by latency.'
})

indexes_df = pd.DataFrame(created_indexes)
indexes_df

,collection,index_name,purpose
0,service_cases,idx_zone_bad_completion,"Supports filtering by pickup zone and bad outcome, with sorting by completion time."
1,service_cases,idx_complaint_type,Supports searches for embedded complaint types inside service cases.
2,service_cases,idx_tracking_gap,Supports fast retrieval of orders that have no linked delivery record.
3,app_events,idx_app_event_latency,Supports filtering app events by event type and sorting by latency.


## **Run Explain Plans After Indexing**

After the indexes are created, the same explain plans are collected again. A good result should show more index usage and fewer documents examined for indexed queries.

In [ ]:
after_summaries = []

for spec in query_specs:
    explain_doc = run_find_explain(
        collection=spec['collection'],
        query=spec['filter'],
        projection=spec['projection'],
        sort=spec['sort'],
        limit=spec['limit']
    )

    summary = summarise_explain(
        query_name=spec['query_name'],
        collection_name=spec['collection'].name,
        explain_doc=explain_doc,
        index_status='after_index'
    )

    after_summaries.append(summary)

    json_path = OUTPUT_PATH / f"explain_after_{safe_query_name(spec['query_name'])}.json"
    save_json(json_path, explain_doc)

after_df = pd.DataFrame(after_summaries)
after_df

,query_name,collection,index_status,n_returned,total_docs_examined,total_keys_examined,execution_time_ms,uses_index,uses_collection_scan,plan_stages
0,Central bad outcome cases,service_cases,after_index,25,25,25,1,True,False,"FETCH, IXSCAN, LIMIT, PROJECTION_DEFAULT"
1,Delay complaint service cases,service_cases,after_index,25,25,25,1,True,False,"FETCH, IXSCAN, LIMIT, PROJECTION_DEFAULT"
2,Tracking gap cases,service_cases,after_index,25,25,25,1,True,False,"FETCH, IXSCAN, LIMIT, PROJECTION_DEFAULT"
3,Delivery instruction update latency,app_events,after_index,25,25,25,1,True,False,"FETCH, IXSCAN, LIMIT, PROJECTION_SIMPLE"


## **Aggregation Pipeline Explain After Indexing**

The same aggregation pipeline is now explained again after the targeted indexes have been created. The aim is to check whether MongoDB can use an index for the `$match` stage before grouping the records.

Even if MongoDB chooses a collection scan on this small dataset, this step still shows the correct optimisation method: define the query pattern, create a relevant index, run explain plans, and compare the result.

In [ ]:
aggregation_explain_after = db.command(
    'explain',
    {
        'aggregate': 'service_cases',
        'pipeline': aggregation_pipeline,
        'cursor': {}
    },
    verbosity='executionStats'
)

aggregation_after_summary = summarise_explain(
    query_name='Aggregation zone bad outcome pipeline',
    collection_name='service_cases',
    explain_doc=aggregation_explain_after,
    index_status='after_index'
)

save_json(OUTPUT_PATH / 'aggregation_explain_after.json', aggregation_explain_after)

aggregation_after_df = pd.DataFrame([aggregation_after_summary])

aggregation_compare_df = aggregation_before_df.merge(
    aggregation_after_df,
    on=['query_name', 'collection'],
    suffixes=('_before', '_after')
)

aggregation_compare_df['docs_examined_reduction_pct'] = aggregation_compare_df.apply(
    lambda row: pct_reduction(row['total_docs_examined_before'], row['total_docs_examined_after']),
    axis=1
)

aggregation_compare_df['keys_examined_change'] = (
    aggregation_compare_df['total_keys_examined_after'].fillna(0)
    - aggregation_compare_df['total_keys_examined_before'].fillna(0)
)

aggregation_compare_df.to_csv(OUTPUT_PATH / 'aggregation_explain_comparison.csv', index=False)

aggregation_compare_df


,query_name,collection,index_status_before,n_returned_before,total_docs_examined_before,total_keys_examined_before,execution_time_ms_before,uses_index_before,uses_collection_scan_before,plan_stages_before,index_status_after,n_returned_after,total_docs_examined_after,total_keys_examined_after,execution_time_ms_after,uses_index_after,uses_collection_scan_after,plan_stages_after,docs_examined_reduction_pct,keys_examined_change
0,Aggregation zone bad outcome pipeline,service_cases,before_index,1250,1250,0,2,False,True,"COLLSCAN, GROUP, filter, group, project, scan",after_index,84,84,84,1,True,False,"FETCH, GROUP, IXSCAN, group, ixseek, limit, nlj, project, seek",93.28,84


## **Compare Before and After Results**

This comparison table checks whether the targeted indexes changed the query execution behaviour. The key metric is `totalDocsExamined`, which shows how many documents MongoDB inspected to answer each query. Lower document examination is better because it means MongoDB is doing less scanning work.

The table also checks whether the explain plan used an index scan or a collection scan before and after the indexes were created.

In [ ]:
comparison_df = before_df.merge(
    after_df,
    on=['query_name', 'collection'],
    suffixes=('_before', '_after')
)

comparison_df['docs_examined_reduction_pct'] = comparison_df.apply(
    lambda row: pct_reduction(row['total_docs_examined_before'], row['total_docs_examined_after']),
    axis=1
)

comparison_df['keys_examined_change'] = (
    comparison_df['total_keys_examined_after'].fillna(0) -
    comparison_df['total_keys_examined_before'].fillna(0)
)

comparison_df['optimisation_result'] = comparison_df.apply(
    lambda row: 'Improved with index' if row['uses_index_after'] else 'Check explain plan',
    axis=1
)

comparison_display = comparison_df[
    [
        'query_name',
        'collection',
        'total_docs_examined_before',
        'total_docs_examined_after',
        'docs_examined_reduction_pct',
        'total_keys_examined_before',
        'total_keys_examined_after',
        'uses_index_before',
        'uses_index_after',
        'uses_collection_scan_before',
        'uses_collection_scan_after',
        'execution_time_ms_before',
        'execution_time_ms_after',
        'optimisation_result'
    ]
]
comparison_display

,query_name,collection,total_docs_examined_before,total_docs_examined_after,docs_examined_reduction_pct,total_keys_examined_before,total_keys_examined_after,uses_index_before,uses_index_after,uses_collection_scan_before,uses_collection_scan_after,execution_time_ms_before,execution_time_ms_after,optimisation_result
0,Central bad outcome cases,service_cases,1250,25,98.00,0,25,False,True,True,False,2,1,Improved with index
1,Delay complaint service cases,service_cases,353,25,92.92,0,25,False,True,True,False,0,1,Improved with index
2,Tracking gap cases,service_cases,124,25,79.84,0,25,False,True,True,False,0,1,Improved with index
3,Delivery instruction update latency,app_events,640,25,96.09,0,25,False,True,True,False,0,1,Improved with index


### **Note on Documents Examined and Small Dataset Behaviour**

`totalDocsExamined` shows how many documents MongoDB had to inspect to answer a query. A high value means the database scanned more records, which can become slower as the dataset grows. After indexing, the preferred result is that MongoDB examines fewer documents or uses index keys to narrow down the search.

The `service_cases` collection contains 1,250 documents, which is small compared with a production database. Because of this, MongoDB may sometimes choose a collection scan even after an index exists, because scanning a small collection can still be inexpensive. This does not automatically mean the index design is wrong. In a larger production system, the same index would be more valuable because it would reduce the number of documents examined for repeated operational queries.

## **Run the Optimised Queries and Save Sample Results**

This step saves small output samples from the optimised queries. These outputs can be used as evidence that the indexed queries still return meaningful business results after optimisation.

In [ ]:
sample_output_frames = {}

for spec in query_specs:
    cursor = spec['collection'].find(
        spec['filter'],
        spec['projection']
    )

    if spec['sort'] is not None:
        cursor = cursor.sort(spec['sort'])

    cursor = cursor.limit(spec['limit'])

    rows = [flatten_dict(doc) for doc in cursor]
    sample_df = pd.DataFrame(rows)

    sample_name = safe_query_name(spec['query_name'])
    sample_output_frames[sample_name] = sample_df

    sample_path = OUTPUT_PATH / f'optimised_query_sample_{sample_name}.csv'
    sample_df.to_csv(sample_path, index=False)

print('Optimised query sample outputs saved:')
for name, sample_df in sample_output_frames.items():
    print(f'{name}: {sample_df.shape[0]} rows')

Optimised query sample outputs saved:
central_bad_outcome_cases: 25 rows
delay_complaint_service_cases: 25 rows
tracking_gap_cases: 25 rows
delivery_instruction_update_latency: 25 rows


## **Inspect Final Indexes**

This step lists the final indexes on the optimised collections. This is useful evidence for the report because it proves the indexes were created in MongoDB Atlas.

In [ ]:
final_index_rows = []

for collection in [service_cases, app_events]:
    for index in collection.list_indexes():
        final_index_rows.append({
            'collection': collection.name,
            'index_name': index.get('name'),
            'key': json.dumps(index.get('key'), default=str)
        })

final_indexes_df = pd.DataFrame(final_index_rows)
final_indexes_df

,collection,index_name,key
0,service_cases,_id_,"{""_id"": 1}"
1,service_cases,idx_zone_bad_completion,"{""zones.pickup"": 1, ""status_flags.is_bad_outcome"": 1, ""analytics.completion_hours_clean"": -1}"
2,service_cases,idx_complaint_type,"{""status_flags.has_complaint"": 1, ""complaints.complaint_type"": 1}"
3,service_cases,idx_tracking_gap,"{""status_flags.tracking_gap_flag"": 1}"
4,app_events,_id_,"{""_id"": 1}"
5,app_events,idx_app_event_latency,"{""event_type"": 1, ""api_latency_ms"": -1}"


## **Save Optimisation Outputs Locally for Coursework Evidence**

In [ ]:
before_df.to_csv(OUTPUT_PATH / 'query_explain_before_summary.csv', index=False)
after_df.to_csv(OUTPUT_PATH / 'query_explain_after_summary.csv', index=False)
comparison_display.to_csv(OUTPUT_PATH / 'query_optimisation_comparison.csv', index=False)
indexes_df.to_csv(OUTPUT_PATH / 'indexes_created.csv', index=False)
final_indexes_df.to_csv(OUTPUT_PATH / 'final_indexes.csv', index=False)
collection_counts_df.to_csv(OUTPUT_PATH / 'query_optimisation_collection_counts.csv', index=False)

print('Query optimisation output files saved locally:')
print(OUTPUT_PATH)
print()

for file_path in sorted(OUTPUT_PATH.iterdir()):
    if file_path.is_file():
        size_kb = file_path.stat().st_size // 1024
        print(f'{file_path.name:<65} {size_kb} KB')

zip_file = '/content/northstar_query_optimisation_outputs.zip'

if Path(zip_file).exists():
    Path(zip_file).unlink()

import shutil
shutil.make_archive('/content/northstar_query_optimisation_outputs', 'zip', OUTPUT_PATH)

print('\nQuery optimisation zip created:')
print(zip_file)
print('\nOptimisation output zip created for coursework evidence.')

Query optimisation output files saved locally:
/content/northstar_query_optimisation_outputs

aggregation_explain_after.json                                    14 KB
aggregation_explain_before.json                                   11 KB
aggregation_explain_comparison.csv                                0 KB
explain_after_central_bad_outcome_cases.json                      6 KB
explain_after_delay_complaint_service_cases.json                  6 KB
explain_after_delivery_instruction_update_latency.json            5 KB
explain_after_tracking_gap_cases.json                             5 KB
explain_before_central_bad_outcome_cases.json                     5 KB
explain_before_delay_complaint_service_cases.json                 4 KB
explain_before_delivery_instruction_update_latency.json           4 KB
explain_before_tracking_gap_cases.json                            3 KB
final_indexes.csv                                                 0 KB
indexes_created.csv                                 

## **Summary of MongoDB Implementation and Query Optimisation**

The MongoDB section created the `NorthStarDB` database, inserted cleaned data into master collections and the `service_cases` collection, demonstrated CRUD operations, and produced aggregation outputs for zone performance, complaints, hub performance, and app latency. The query optimisation section then evaluated key business queries with explain plans, created targeted indexes, compared before and after execution statistics, and added an aggregation pipeline explain example for stronger optimisation evidence.